<a href="https://colab.research.google.com/github/YouCanKeepSilence/ml_labs/blob/master/ML_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Импорт данных

In [0]:
import numpy as np
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Загружаем данные из .csv

In [0]:
names_array = ['gender', 'age', 'first_label', 'second_label', 'third_label', *['c_{}'.format(i + 1) for i in range(60000)]]
url = '/content/drive/My Drive/ML/half.csv'
data = pd.read_csv(url, names=names_array)

# Обработка данных

Удаляем ненужные стобцы (по причине малого количества данных в них)

In [0]:
data = data.drop(['second_label', 'third_label'], axis=1)

In [5]:
#new_data = data # to skip augmentation because it'll break PCA method
#new_batch = 5000
#sensors_count = 12
data

gender  age  first_label  ...   c_59998   c_59999   c_60000
0     Female   85            3  ... -0.042003 -0.034466 -0.039000
1       Male   79            9  ... -0.819597 -0.771195 -0.845490
2       Male   66            3  ...  0.030947  0.034947  0.037947
3       Male   93            2  ... -0.016540 -0.016289 -0.022542
4       Male   65            2  ... -0.026058 -0.012157 -0.001676
...      ...  ...          ...  ...       ...       ...       ...
1245    Male   74            2  ...  0.675640  0.669640  0.661640
1246  Female   85            7  ... -0.053092 -0.049083 -0.049954
1247    Male   61            2  ...  0.072895  0.062235  0.060652
1248  Female   85            3  ... -0.032233 -0.030831 -0.033880
1249    Male   82            1  ... -0.035582 -0.215320 -0.406685

[1250 rows x 60003 columns]

## Агументация данных

In [6]:
# batch_size = 1000
def augmentate():
  measurment_names = data.columns[3:]
  new_data_inner = pd.DataFrame()
  # Пользуемся тем, что имеем 5к измерений для каждого из 12 датчиков. Соответственно нарежем по 250 с каждого из 12 датчиков
  new_batch = 250
  measurments_for_sensors = 5000
  sensors_count = 12
  for i in range(0, measurments_for_sensors, new_batch):
      current_measurment_names = {f'c_{k + j * measurments_for_sensors + 1}': f'c_{(new_batch * j + k % new_batch) % (new_batch * sensors_count)  + 1}' for j in range(sensors_count) for k in range(i, i + new_batch)}
      mapped_piece = data[['age', 'gender', 'first_label', *[f'c_{k + j * measurments_for_sensors + 1}'  for j in range(sensors_count) for k in range(i, i + new_batch)]]].rename(columns=current_measurment_names)
      new_data_inner = pd.concat([new_data_inner, mapped_piece])
  return new_data_inner

# don't call augmentate...
new_data = augmentate()

age  gender  first_label  ...    c_2998    c_2999    c_3000
0      85  Female            3  ...  0.157333  0.168144  0.167612
1      79    Male            9  ...  0.086233  0.073918  0.061897
2      66    Male            3  ... -0.032053 -0.032053 -0.032053
3      93    Male            2  ... -0.077995 -0.083645 -0.090060
4      65    Male            2  ... -0.067565 -0.061280 -0.054876
...   ...     ...          ...  ...       ...       ...       ...
1245   74    Male            2  ...  0.675640  0.669640  0.661640
1246   85  Female            7  ... -0.053092 -0.049083 -0.049954
1247   61    Male            2  ...  0.072895  0.062235  0.060652
1248   85  Female            3  ... -0.032233 -0.030831 -0.033880
1249   82    Male            1  ... -0.035582 -0.215320 -0.406685

[25000 rows x 3003 columns]

Таким образом получаем из размерности 1250, 60003 -> 25000, 3003 увеличив количество записей в датасете в 20 раз



## Обработаем пропущенные значения

In [7]:
print(new_data['age'].unique())
print(new_data['gender'].unique())
print(new_data['first_label'].unique())
print(new_data.isnull().sum())

[85 79 66 93 65 55 82 87 51 61 90 77 70 60 62 59 84 71 45 48 75 49 72 78
 86 46 94 33 74 64 67 63 88 32 92 68 76 54 89 31 80 30 57 83 34 27 73 69
 52 44 43 47 39 56 81 53 50 16 29 37 58 24 22 40 41 11 26 28 96 35 91 25
 36 23 19 17 42 18 15  8 20 38 14 21]
['Female' 'Male']
[3 9 2 5 1 7 6 4 8]
age            0
gender         0
first_label    0
c_1            0
c_2            0
              ..
c_2996         0
c_2997         0
c_2998         0
c_2999         0
c_3000         0
Length: 3003, dtype: int64


Пропущенных значений нет.

## Обработаем категориальные признаки

In [8]:
categorial_columns = ['gender', 'first_label']
numerical_columns = ['age', *[f'c_{i + 1}' for i in range(new_batch * sensors_count)]]
print(categorial_columns)

['gender', 'first_label']


Проведем бинаризацию пола, label не трогаем, им займемся дальше

In [9]:
new_data.at[data['gender'] == 'Male', 'gender'] = 0
new_data.at[data['gender'] == 'Female', 'gender'] = 1
new_data['gender'].describe()

count    25000.000000
mean         0.392800
std          0.488383
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          1.000000
Name: gender, dtype: float64

In [0]:
categorial_columns.remove('gender')
numerical_columns.append('gender')

## Произведем нормировку данных

In [11]:
numerical_data = new_data[numerical_columns]
numerical_data.describe()

age           c_1  ...        c_3000        gender
count  25000.000000  25000.000000  ...  25000.000000  25000.000000
mean      65.856000      0.000762  ...     -0.001046      0.392800
std       16.253524      0.136387  ...      0.290090      0.488383
min        8.000000     -2.098990  ...     -4.676025      0.000000
25%       58.000000     -0.043578  ...     -0.080752      0.000000
50%       68.000000     -0.013588  ...     -0.031496      0.000000
75%       78.000000      0.019670  ...      0.027639      1.000000
max       96.000000      5.972125  ...      7.605975      1.000000

[8 rows x 3002 columns]

In [0]:
mean = numerical_data.mean(axis = 0)
std = numerical_data.std(axis = 0)
numerical_data = (numerical_data - mean) / std
new_data[numerical_columns] = numerical_data

In [13]:
new_data.describe()

age        gender  ...        c_2999        c_3000
count  2.500000e+04  2.500000e+04  ...  2.500000e+04  2.500000e+04
mean   3.352074e-16 -6.536993e-17  ...  4.416467e-17  1.271871e-17
std    1.000000e+00  1.000000e+00  ...  1.000000e+00  1.000000e+00
min   -3.559597e+00 -8.042872e-01  ... -1.622698e+01 -1.611564e+01
25%   -4.833413e-01 -8.042872e-01  ... -2.757850e-01 -2.747627e-01
50%    1.319099e-01 -8.042872e-01  ... -1.056542e-01 -1.049655e-01
75%    7.471611e-01  1.243287e+00  ...  9.621792e-02  9.888532e-02
max    1.854613e+00  1.243287e+00  ...  2.629583e+01  2.622300e+01

[8 rows x 3003 columns]

Нормализовали все поля кроме label т.к. он является выходным и его нет смысла нормализовывать.

# Подготовка данных для модели

In [14]:
from sklearn.model_selection import train_test_split
y = new_data['first_label'].values
X = new_data.drop(['first_label'], axis=1).values

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

n_train, _ = x_train.shape 
n_test,  _ = x_test.shape 

print('Train dataset size: {}'.format(n_train))
print('Test dataset size: {}'.format(n_test))

Train dataset size: 17500
Test dataset size: 7500


# Создание и обучение моделей

## KMeans

In [0]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=len(new_data['first_label'].unique()), random_state=42).fit(X)

In [16]:
kmeans_cluster = kmeans.predict(X) + 1
print(np.unique(kmeans.labels_))
check = kmeans_cluster == y
np.where(check == True)[0].shape

[0 1 2 3 4 5 6 7 8]


(3577,)

Т.к. разбиение подразумевает передачу количества классов, то количество классов совпадает, однако пересечение с реальным разбиением достаточно слабое.

## DBSCAN

In [0]:
from sklearn.cluster import DBSCAN
dbscan = DBSCAN().fit(X)

In [18]:
np.unique(dbscan.labels_)

array([-1,  0,  1,  2,  3])

С таким набором параметров разбиение происходит на 5 классов, в то время как изначальное разбиение произведено на 9 классов

## AffinityPropagation
Не удалось проверить ибо вылезло за пределы по памяти (более 25 Гб оперативки)

In [0]:
from sklearn.cluster import AffinityPropagation
affinity = AffinityPropagation().fit(X)

In [0]:
np.unique(affinity.labels_)

## Spectral Clustering
Идет очень долго, более часа... Неизвестно закончит ли

In [19]:
from sklearn.cluster import SpectralClustering
spectral = SpectralClustering(n_clusters=len(new_data['first_label'].unique()), random_state=42).fit(X)

/usr/local/lib/python3.6/dist-packages/sklearn/manifold/_spectral_embedding.py:236: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.
  warnings.warn("Graph is not fully connected, spectral embedding"


KeyboardInterrupt: ignored

In [0]:
spectral_cluster = spectral.predict(X) + 1
print(np.unique(spectral.labels_))
check_spectral = spectral_cluster == y
np.where(check_spectral == True)[0].shape

## Birch
с 19:03

In [0]:
from sklearn.cluster import Birch
birch = Birch(n_clusters=len(new_data['first_label'].unique())).fit(X)

In [0]:
birch_cluster = birch.predict(X) + 1
print(np.unique(birch.labels_))
check_birch = birch_cluster == y
np.where(check_birch == True)[0].shape

# Вывод
